src/gas_metering/mod.rs

```diff
@@ -284,7 +284,105 @@ struct MeteredBlock {
 	/// Index of the first instruction (aka `Opcode`) in the block.
 	start_pos: usize,
 	/// Sum of costs of all instructions until end of the block.
-	cost: u32,
+	cost: BlockCostCounter,
+}
+
+/// Metering block cost counter, which handles arithmetic overflows.
+///
+/// Arithmetical overflows can occur while summarizing costs of some
+/// instruction set. To handle this, we count amount of such overflows
+/// with a separate counter and continue counting cost of metering block.
+///
+/// Obviously, overflows counter can overflow as well. A solution here
+/// could be usage of `Vec` each element of which is a separate counter
+/// of overflows. Once last overflow counter overflowed itself, a new is
+/// pushed to the vector and starts again. But this solution can be optimized
+/// after recognizing a property of this vector: all the elements before
+/// the last one are equal to `u32::MAX`, i.e. they all are overflowed
+/// counters of gas calculation overflows. So we can optimize the solution
+/// the next way: to avoid using `Vec` (redundant allocations on the heap)
+/// we can divide the vector into two fields:
+/// 1. Length of the vector's slice, each element of which has maximum
+/// value (`u32::MAX`). This field explicitly shows property we stated
+/// above.
+/// 2. Current overflows counter value, which is not overflowed yet.
+/// This field has the same semantics as the last element of the vector
+/// described earlier.
+///
+/// The latter optimization approach is implemented and 2 fields are introduced:
+/// `max_out_overflows_num` and `overflows`.
+#[derive(Debug, PartialEq, PartialOrd)]
+#[cfg_attr(test, derive(Copy, Clone, Default))]
+struct BlockCostCounter {
+	/// Amount of overflow counters that overflowed themselves.
+	/// For more info see type docs.
+	max_out_overflows_num: usize,
+	/// Current amount of overflows.
```

Comment:
Why we cant just use `u64` here instead all of this fields. Then in gas injection we will split u64 into two parts [u32, u32] and then we can use it just like all this fields

Reply:
discussed